todo_app/app.py

```diff
@@ -3,69 +3,168 @@
+    def unauthenticated():  
+        app.logger.info(f"Session is Unauthenticated!")
+        client = WebApplicationClient(client_id)
+        redirect_uri=f'{request.url}login/callback'
+        uri = client.prepare_request_uri('https://github.com/login/oauth/authorize', redirect_uri=redirect_uri)
+        app.logger.info(f"Authentication Step 1) Ask GitHub for Authorization Code")
+        app.logger.info(f"Ask GitHub to Redirect to: {redirect_uri}")
+        app.logger.info(f"Redirecting to: {uri}")
+        return redirect(uri)
+    
+    @login_manager.user_loader 
+    def load_user(user_id): 
+        return User(user_id)         
 
     @app.route('/')
+    @login_required
     def index():
         try:
-            app.logger.info("index()")
+            username = current_user.get_id()
+            app.logger.info(f"[{username}] index() => loading tasks from repository")
             tasks = repository.get_tasks()
-            app.logger.info(f"index() => {len(tasks)} tasks retrieved from repository")
-            view_model = ViewModel(tasks)
+            app.logger.info(f"[{username}] index() => {len(tasks)} tasks retrieved from repository")
+            view_model = ViewModel(tasks, username, current_user.get_role())
             response = render_template('index.html', view_model=view_model) 
             return response
         except Exception as e:
             app.logger.error(f"Error : {e}")
             raise e
 
-    @app.route('/tasks', methods=['POST'])
-    def add_new_task():
-        try:
-            app.logger.info("add_new_task()")
-            task_name = request.form["task_name"]
-            app.logger.info(f"add_new_task() => {task_name}")
-            repository.add_task(task_name, "ToDo")
+    @app.route('/login/callback', methods=['GET'])
+    def login_callback():
+        try:            
+            app.logger.info(f"Authentication Step 1) Completed.")
+            auth_code = request.args["code"]            
+            app.logger.info(f"GitHub Authorization Code: {auth_code}")
+
+            # exchange the authorization code for an access token
+            payload = {"client_id": client_id, "client_secret": client_secret, "code": auth_code}
+            headers = {"Accept": "application/json"}
+            app.logger.info(f"Authentication Step 2) Exchange GitHub Authorization Code for Access Token")
+            app.logger.info(f"POST: https://github.com/login/oauth/access_token")
+            app.logger.info(f"Payload: {payload}")
+            r = requests.post("https://github.com/login/oauth/access_token", data = payload, headers = headers)
+            app.logger.info(f"Authentication Step 2) Complete. Response: {r.text}")
+
+            if "access_token" not in r.json():
+                # HELP - is there a better way to handle this? how to return 401 - Unauthorized?
+                raise Exception("Authentication Failed at Step 2. See application logs.")
```

Comment:
You can provide a specific status code with Flask by including it as the second argument for a tuple in the return statement, (e.g. [see this SO discussion](https://stackoverflow.com/questions/7824101/return-http-status-code-201-in-flask)) so for example:
```suggestion
                return "Authentication Failed at Step 2. See application logs.", 401
```